### ISSUE TITLE
Conversion errors with `SQLAlchemy==2.0.2`

### ISSUE THREAD

**Issue by kqf** [marshmallow-code/marshmallow-sqlalchemy#487]
## Description
It looks like `ma-sqla` doesn't work with the newest version of `SQLAlchemy==2.0.2`. Here is an example taken from docs compiled into a single file:

```python
import sqlalchemy as sa
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import scoped_session, sessionmaker, relationship, backref
from marshmallow_sqlalchemy import SQLAlchemySchema, auto_field

engine = sa.create_engine("sqlite:///:memory:")
session = scoped_session(sessionmaker(bind=engine))
Base = declarative_base()


class Author(Base):
    __tablename__ = "authors"
    id = sa.Column(sa.Integer, primary_key=True)
    name = sa.Column(sa.String, nullable=False)

    def __repr__(self):
        return "<Author(name={self.name!r})>".format(self=self)


class Book(Base):
    __tablename__ = "books"
    id = sa.Column(sa.Integer, primary_key=True)
    title = sa.Column(sa.String)
    author_id = sa.Column(sa.Integer, sa.ForeignKey("authors.id"))
    author = relationship("Author", backref=backref("books"))


Base.metadata.create_all(engine)


class AuthorSchema(SQLAlchemySchema):
    class Meta:
        model = Author
        load_instance = True  # Optional: deserialize to model instances

    id = auto_field()
    name = auto_field()
    books = auto_field()


class BookSchema(SQLAlchemySchema):
    class Meta:
        model = Book
        load_instance = True

    id = auto_field()
    title = auto_field()
    author_id = auto_field()
```

When I am using `SQLAlchemy==2.0.1` everything works, with a minor warning. But when I am update to `SQLAlchemy==2.0.2` I get the following error: 

```
../test.py:8: MovedIn20Warning: The ``declarative_base()`` function is now available as sqlalchemy.orm.declarative_base(). (deprecated since: 2.0) (Background on SQLAlchemy 2.0 at: https://sqlalche.me/e/b8d9)
  Base = declarative_base()
Traceback (most recent call last):
  File "../test.py", line 31, in <module>
    class AuthorSchema(SQLAlchemySchema):
  File ".pyenv/versions/elan/lib/python3.10/site-packages/marshmallow/schema.py", line 121, in __new__
    klass._declared_fields = mcs.get_declared_fields(
  File ".pyenv/versions/elan/lib/python3.10/site-packages/marshmallow_sqlalchemy/schema.py", line 92, in get_declared_fields
    fields.update(mcs.get_auto_fields(fields, converter, opts, dict_cls))
  File ".pyenv/versions/elan/lib/python3.10/site-packages/marshmallow_sqlalchemy/schema.py", line 102, in get_auto_fields
    {
  File ".pyenv/versions/elan/lib/python3.10/site-packages/marshmallow_sqlalchemy/schema.py", line 103, in <dictcomp>
    field_name: field.create_field(
  File ".pyenv/versions/elan/lib/python3.10/site-packages/marshmallow_sqlalchemy/schema.py", line 28, in create_field
    return converter.field_for(model, column_name, **self.field_kwargs)
  File ".pyenv/versions/elan/lib/python3.10/site-packages/marshmallow_sqlalchemy/convert.py", line 222, in field_for
    attr = getattr(model, property_name)
AttributeError: type object 'Author' has no attribute 'books'
```
It looks like heuristics in `marshmallow_sqlalchemy/convert.py` don't correspond to the `SQLAlchemy` in the new version. I noticed it for my personal use case (I am not showing it here) fails to detect [relationship during the conversion](https://github.com/marshmallow-code/marshmallow-sqlalchemy/blob/dev/src/marshmallow_sqlalchemy/convert.py#L272) 

## Env
It was spotted in `python:3.10` the requirements are:
```
marshmallow-sqlalchemy==0.28.1
SQLAlchemy==2.0.2
```

**Comment by lafrech** [comment]
I'm afraid SQLA 2 is not supported at all (CI fails with 2.0). See #488.

**Comment by kqf** [comment]
Oh, then it was a pure luck that it worked for my setup 😅 

@lafrech many thanks for taking initiative for that. Je te souhaite bonne chance 🤞

**Comment by rhynzler** [comment]
The problem is because in the `fields_for_model` method of the `ModelConverter` class iterates over `_orm.Mapper.iterate_properties` which no longer implicitly calls `_orm.registry.configure()` in version 2.0.2. A temporary solution could be to modify the iteration so that it goes through `_orm.Mapper.attrs` which does implicitly call `_orm.registry.configure()`

Ref: https://docs.sqlalchemy.org/en/20/changelog/changelog_20.html#change-45fe5664c88893d899c991ca3675142c

**Comment by doyond** [comment]
Rhynzler above is right. 

I have faced also issues with relationships due to this. For anyone else facing this issue, for now my cleanest option has been to call configure_mappers() before importing my schemas.

`from sqlalchemy.orm import configure_mappers`
`configure_mappers()`

**Comment by YuvalItzchakov** [comment]
Ran into the same issue and [wrote this SO question](https://stackoverflow.com/q/75457741/1870803). I tried using `configure_mappers()` before loading my models as well:

```python
from sqlalchemy.orm import configure_mappers
configure_mappers()
from models import *
from views import *
```

This however did not work and still gave me errors when trying to derive marshmallow-sqlalchemy schemas.
However I was successfull in explicitly iterating the relationship objects which I believe may cause the same side effect:

```python
mapper = inspect(cls)
_ = [_ for _ in mapper.relationships]
```

**Comment by doyond** [comment]
> Ran into the same issue and [wrote this SO question](https://stackoverflow.com/q/75457741/1870803). I tried using `configure_mappers()` before loading my models as well:
> 
> ```python
> from sqlalchemy.orm import configure_mappers
> configure_mappers()
> from models import *
> from views import *
> ```
> 
> This however did not work and still gave me errors when trying to derive marshmallow-sqlalchemy schemas. However I was successfull in explicitly iterating the relationship objects which I believe may cause the same side effect:
> 
> ```python
> mapper = inspect(cls)
> _ = [_ for _ in mapper.relationships]
> ```

The configure_mappers() should be called **after** importing SQLAlchemy models, where the relationships are defined, but **before** the Schemas are imported.

In my particular setup, I have a models folder and a schema folders and things get initialized in that order:

models/
__init__.py -> this will import all the models from each of the models file, I have ~ 1 file per table. Once the models are imported, it will import schemas. (see init below)

schemas/
__init__.py -> This will first configure mappers at the top, then import all the schemas from their respective files.

Regardless if your file structure, calling configure_mappers() before the models are imported will not work.

**Comment by schalkje** [comment]
Splitting model from schema's and calling configure_mappers(); worked for me

**Comment by LloydCapson** [comment]
I've got a pull request in that fixes this issue. [https://github.com/marshmallow-code/marshmallow-sqlalchemy/pull/491](url)

**Comment by lafrech** [comment]
Thanks @LloydCapson. Please rebase your PR. Also please reword the comment (it sounds like we're blaming SQLA maintainers). Anyone here, #491 PR review welcome.

Thanks!

**Comment by LloydCapson** [comment]
Ok I believe I've got it rebased correctly and cleaned up. Let me know if there's anything wrong with it and I'll fix it. I'm making an effort to learn to contribute.

**Comment by YuvalItzchakov** [comment]
> > Ran into the same issue and [wrote this SO question](https://stackoverflow.com/q/75457741/1870803). I tried using `configure_mappers()` before loading my models as well:
> > 
> > ```python
> > from sqlalchemy.orm import configure_mappers
> > configure_mappers()
> > from models import *
> > from views import *
> > ```
> > 
> > This however did not work and still gave me errors when trying to derive marshmallow-sqlalchemy schemas. However I was successfull in explicitly iterating the relationship objects which I believe may cause the same side effect:
> > 
> > ```python
> > mapper = inspect(cls)
> > _ = [_ for _ in mapper.relationships]
> > ```
> 
> The configure_mappers() should be called **after** importing SQLAlchemy models, where the relationships are defined, but **before** the Schemas are imported.
> 
> In my particular setup, I have a models folder and a schema folders and things get initialized in that order:
> 
> models/
> __init__.py -> this will import all the models from each of the models file, I have ~ 1 file per table. Once the models are imported, it will import schemas. (see init below)
> 
> schemas/
> __init__.py -> This will first configure mappers at the top, then import all the schemas from their respective files.
> 
> Regardless if your file structure, calling configure_mappers() before the models are imported will not work.

The issue I have with that approach is that my models have a decorator to them that automagically derives the marshmallow schema. I can derive all models at the end by iterating them all, would just be a bit less convenient.

**Comment by doyond** [comment]
> > > Ran into the same issue and [wrote this SO question](https://stackoverflow.com/q/75457741/1870803). I tried using `configure_mappers()` before loading my models as well:
> > > ```python
> > > from sqlalchemy.orm import configure_mappers
> > > configure_mappers()
> > > from models import *
> > > from views import *
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > This however did not work and still gave me errors when trying to derive marshmallow-sqlalchemy schemas. However I was successfull in explicitly iterating the relationship objects which I believe may cause the same side effect:
> > > ```python
> > > mapper = inspect(cls)
> > > _ = [_ for _ in mapper.relationships]
> > > ```
> > 
> > 
> > The configure_mappers() should be called **after** importing SQLAlchemy models, where the relationships are defined, but **before** the Schemas are imported.
> > In my particular setup, I have a models folder and a schema folders and things get initialized in that order:
> > models/
> > **init**.py -> this will import all the models from each of the models file, I have ~ 1 file per table. Once the models are imported, it will import schemas. (see init below)
> > schemas/
> > **init**.py -> This will first configure mappers at the top, then import all the schemas from their respective files.
> > Regardless if your file structure, calling configure_mappers() before the models are imported will not work.
> 
> The issue I have with that approach is that my models have a decorator to them that automagically derives the marshmallow schema. I can derive all models at the end by iterating them all, would just be a bit less convenient.

I see. On my side I avoid using the decorators because I almost always need to do something custom at the schema level. I personally use the SQLAlchemyAutoSchema to "automatically" generate the schemas, although it still requires a few lines of code. However, I think those can be more easily extended with additional functionalities.

Something like:
https://marshmallow-sqlalchemy.readthedocs.io/en/latest/

If ever you are still not happy with your current solution and can't wait for the next version of marshmallow, you could try something similar. The footprint is larger than a simple decorator, but I think there are also other benefits... Either way, if your solution works, that's fine too. Eventually Marshmallow will just be patched.

**Comment by lafrech** [comment]
@rhynzler's [comment](https://github.com/marshmallow-code/marshmallow-sqlalchemy/issues/487#issuecomment-1424568525) points to SQLAlchemy's changelog which seems to indicate that we'd rather be using `Mapper.attrs`. I have no time right now to investigate this but would @LloydCapson or anyone be willing to try this?

**Comment by doyond** [comment]
> @rhynzler's [comment](https://github.com/marshmallow-code/marshmallow-sqlalchemy/issues/487#issuecomment-1424568525) points to SQLAlchemy's changelog which seems to indicate that we'd rather be using `Mapper.attrs`. I have no time right now to investigate this but would @LloydCapson or anyone be willing to try this?

I've tried this before, it works fine as it does an implicit call to registry.configure(). However, it may be more a patch than anything, and may break again in the future. That being said, it definitely works.

It is equivalent to my suggestion above of calling configure_mappers(), but implicitly. For Marshmallow users, that is definitely a better patch. We do need to keep in mind that  if SQLAlchemy phased that one out, we may eventually run into more issues, but we can kick the bucket down the road.

I can submit a PR for this if we want to go along with it. It's a very minor modification.

**Comment by doyond** [comment]
@lafrech @LloydCapson 

Swapping the model.__mapper__.iterate_properties for model.__mapper__.attrs works fine, and is equivalent to how it worked before. We do not need to explicitly make a call to _check_configure() as in the PR currently presented.

That being said, while most tests are now passing -- we are still getting some tests failures that should be investigated, which are the same whether we use _check_configure() or the __mapper__.attrs:

`FAILED tests/test_conversion.py::TestFieldFor::test_field_for - AttributeError: columns
ERROR tests/test_sqlalchemy_schema.py::test_load[sqla_schema_with_relationships] - AttributeError: columns
ERROR tests/test_sqlalchemy_schema.py::TestLoadInstancePerSchemaInstance::test_toggle_load_instance_per_schema[schema_no_load_instance] - AttributeError: columns
ERROR tests/test_sqlalchemy_schema.py::TestLoadInstancePerSchemaInstance::test_toggle_load_instance_per_schema[schema_with_load_instance] - AttributeError: columns
ERROR tests/test_sqlalchemy_schema.py::test_load_validation_errors[sqla_schema_with_relationships] - AttributeError: columns`

**Comment by lafrech** [comment]
I guess the original sin was to use an underscore method. This broke in a SQLAlchemy patch release. (Luckily, this was on SQLAlchemy 2.0 which is not supported yet anyway.)

Let's not make the same mistake twice. If it is equivalent to use `.attrs` as advised in SQLAlchemy's changelog, go for it.

Thank you guys for investigating this.

**Comment by doyond** [comment]
@lafrech 

The following https://github.com/marshmallow-code/marshmallow-sqlalchemy/pull/495 fixes the issue with the lack of implicit calls to registry.configure() for both iterate_properties and get_property. It now passes all tests.

Feel free to take a look -- and also unless there are any other compatibility issues with SQLAlchemy 2.0, perhaps we can also adjust the requirements as well. I haven't tested anything beyond the unit tests already available.

**Comment by lafrech** [comment]
Thanks @ddoyon92.

I was wrong in my comment: the method we used was not underscored. It's strange that it was removed but there must be a rationale for it. Anyway, following the advice in the changelog and using `.attrs` is the way to go.

I'll review and merge.

We may want to double-check the migration guide before finalizing a SQlAlchemy 2.x compatible release.

**Comment by lafrech** [comment]
Tests pass on SQLAlchemy 2.0. I also tested in my code base.

Would you guys like to test in your code bases before we ship?

**Comment by doyond** [comment]
I've tested against our codebase, everything works as expected and our test suite is passing. I think this particular change is fairly safe.

**Comment by lafrech** [comment]
Yes I also think the change in #495 is safe.

My point is to test your code base against against SQLAlchemy 2.0 (with latest version in dev here) before we ship the new version with official SQLALchemy 2.0 support.

**Comment by doyond** [comment]
Ah I see. We've already migrated to SQLAlchemy 2.0.4 --> I had already implemented the configure_mapper() call in our codebase before importing the schemas, and that was our only problem encountered during migration. I simply removed that patch on our side, and tested the latest dev version instead. That worked fine.

For our use case, we did not encounter any other issues with SQLAlchemy 2.0+ except for that one which we addressed.
